memgpt/schemas/message.py

```diff
@@ -0,0 +1,489 @@
+import json
+import uuid
+from datetime import datetime, timezone
+from typing import List, Optional
+
+from pydantic import Field, field_validator
+
+from memgpt.constants import JSON_ENSURE_ASCII, TOOL_CALL_ID_MAX_LEN
+from memgpt.local_llm.constants import INNER_THOUGHTS_KWARG
+from memgpt.schemas.memgpt_base import MemGPTBase
+from memgpt.schemas.openai.chat_completions import ToolCall
+from memgpt.utils import is_utc_datetime
+
+
+class Message(MemGPTBase):
+    """
+    Representation of a message sent.
+
+    Messages can be:
+    - agent->user (role=='agent')
+    - user->agent and system->agent (role=='user')
+    - or function/tool call returns (role=='function'/'tool').
+    """
+
+    __id_prefix__ = "message"
+    id: str = MemGPTBase.generate_id_field()
+    role: str = Field(..., description="The role of the participant.")
+    text: str = Field(..., description="The text of the message.")
+    user_id: str = Field(None, description="The unique identifier of the user.")
+    agent_id: str = Field(None, description="The unique identifier of the agent.")
+    model: Optional[str] = Field(None, description="The model used to make the function call.")
+    name: Optional[str] = Field(None, description="The name of the participant.")
+    created_at: Optional[datetime] = Field(None, description="The time the message was created.")
+    tool_calls: Optional[List[ToolCall]] = Field(None, description="The list of tool calls requested.")
+    tool_call_id: Optional[str] = Field(None, description="The id of the tool call.")
+
+    @field_validator("role")
+    @classmethod
+    def validate_role(cls, v: str) -> str:
+        roles = ["system", "assistant", "user", "tool"]
+        assert v in roles, f"Role must be one of {roles}"
+        return v
```

Comment:
Would it make sense to use either an enum or a literal typing here 

```Python

# in an enums module
class MessageRole(str, Enum):
    assistant = "assistant"
    user = "user"
    tool = "tool"


class Message(BaseModel)
...
role: MessageRole 

```

That dries these out nicely so you never need to check roles outside of typing them